# DATA CLEANING + EDA (Pandas cont.)

Created By: Angelica Rojas

In [1]:
import pandas as pd
import re

## Upload Data

The data for this notebook could be found at this link: https://data.cityofberkeley.info/Public-Safety/Berkeley-PD-Calls-for-Service/k2nh-s5h5 . 

For the purpose of this lesson we will use the CSV file of the data.

In [2]:
df = pd.read_csv("BerkeleyPD_Calls_for_Service.csv")

#what does this do?
df.head()

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

             CVLEGEND  CVDOW                InDbDate  \
0  BURGLARY - VEHICLE      3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE      5  09/25/2017 03:30:18 AM   
2             LARCENY      6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE      0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION      6  09/25/2017 03:30:14 AM   

                                      Block_Location  \
0  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  
2               2500 SHATTUCK AVE  Berkeley    CA  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA

Why did we only want to display the first 5 rows of the dataframe?

What if we wanted to see the size of this dataframe?

In [3]:
# number of rows
len(df.index)

5617

In [5]:
# shape of df (rows, columns)

df.shape

(5617, 11)

# Part 1: DATA CLEANING

## Column Names

What do all these column names even mean? 

On that same website, BPD offers a narrative pdf file that describes the data they provided. https://data.cityofberkeley.info/Public-Safety/Berkeley-PD-Calls-for-Service/k2nh-s5h5


<img src = "DF_col_desc.png">

## Change Column Name(s)

Why would we want to change the column names?

In [6]:
#df = df.rename(columns={'ORIG_COL_NAME': 'NEW_COL_NAME'})
df = df.rename(columns={'CVLEGEND': 'EVENTDESC', 'CVDOW':"D.O.W."})
df.head()

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

            EVENTDESC  D.O.W.                InDbDate  \
0  BURGLARY - VEHICLE       3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE       5  09/25/2017 03:30:18 AM   
2             LARCENY       6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE       0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION       6  09/25/2017 03:30:14 AM   

                                      Block_Location  \
0  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  
2               2500 SHATTUCK AVE  Berkeley    CA  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA

## Investigating Columns

What is the difference between "Block_Location" and "BLKADDR" columns in the dataframe? From the look of the DF that is displayed it looks as though they are almost the same.

In [7]:
#Let's look at the first value in "Block_Location"
df["Block_Location"][0]

'ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n(37.869363, -122.268028)'

In [8]:
#Let's look at the first value in "BLKADDR"
df["BLKADDR"][0]

'ALLSTON WAY & SHATTUCK AVE'

## Create New Columns

What is new information about "Block_Location" that we can actually use and save?

Let's create new columns for the information we extracted from those values.

In [9]:
coordinates = [["".join(x.split()) for x in re.split(r'[()]',i) if x.strip()][-1] for i in df["Block_Location"]]

#new values
longitude =[["".join(x.split()) for x in re.split(r'[,]',i) if x.strip()][-1] for i in coordinates]
latitude = [["".join(x.split()) for x in re.split(r'[,]',i) if x.strip()][0] for i in coordinates]

#create new columns for latitude and longitude
df["LATITUDE"] = latitude
df["LONGITUDE"] = longitude

#Check if it worked
df.head()



CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

            EVENTDESC  D.O.W.                InDbDate  \
0  BURGLARY - VEHICLE       3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE       5  09/25/2017 03:30:18 AM   
2             LARCENY       6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE       0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION       6  09/25/2017 03:30:14 AM   

                                      Block_Location  \
0  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  37.871884  -122.270752  
2               2500 SHATTUCK AVE  Berkeley    CA  37.863811  -122.267412  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  37.867221  -122.263531  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA  37.871884  -122.270752

In [43]:
df.LATITUDE.unique()

array(['37.869363', '37.871884', '37.863811', ..., '37.877247',
       '37.874537', '37.861541'], dtype=object)

In [38]:
df.LATITUDE
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5617 entries, 0 to 5616
Data columns (total 13 columns):
CASENO            5617 non-null int64
OFFENSE           5617 non-null object
EVENTDT           5617 non-null object
EVENTTM           5617 non-null object
EVENTDESC         5617 non-null object
D.O.W.            5617 non-null int64
InDbDate          5617 non-null object
Block_Location    5617 non-null object
BLKADDR           5590 non-null object
City              5617 non-null object
State             5617 non-null object
LATITUDE          5617 non-null object
LONGITUDE         5617 non-null object
dtypes: int64(2), object(11)
memory usage: 570.5+ KB


In [44]:
df2 = df[df['LATITUDE'].str.startswith('37')]

In [57]:
df2.LATITUDE

0       37.869363
1       37.871884
2       37.863811
3       37.867221
4       37.871884
5       37.868706
6       37.865849
7       37.880266
8       37.882457
9       37.856195
10      37.879708
11      37.859557
12      37.868352
13      37.821533
14      37.892152
15      37.878644
16       37.87091
17      37.887498
18      37.861672
19      37.858214
20      37.853275
22       37.86939
23      37.858165
24      37.859557
25       37.88015
26      37.880262
27      37.858628
28      37.880227
29      37.867852
30      37.861604
          ...    
5585    37.855435
5586    37.881141
5587    37.874489
5588     37.87247
5589    37.857409
5590    37.863072
5591    37.866426
5592    37.875531
5593     37.84827
5594      37.8704
5595    37.867212
5596    37.863839
5597    37.870417
5598    37.869921
5599    37.857099
5600    37.865711
5601    37.870289
5602    37.888784
5603    37.864826
5604    37.864535
5605    37.857674
5607    37.868512
5608    37.859309
5610    37.865816
5611    37

In [50]:
df.LONGITUDE.unique()

array(['-122.268028', '-122.270752', '-122.267412', ..., '-122.27708',
       '-122.263862', '-122.251156'], dtype=object)

In [49]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5617 entries, 0 to 5616
Data columns (total 13 columns):
CASENO            5617 non-null int64
OFFENSE           5617 non-null object
EVENTDT           5617 non-null object
EVENTTM           5617 non-null object
EVENTDESC         5617 non-null object
D.O.W.            5617 non-null int64
InDbDate          5617 non-null object
Block_Location    5617 non-null object
BLKADDR           5590 non-null object
City              5617 non-null object
State             5617 non-null object
LATITUDE          5617 non-null object
LONGITUDE         5617 non-null object
dtypes: int64(2), object(11)
memory usage: 570.5+ KB


In [58]:
df2 = df[df['LONGITUDE'].str.startswith('-122')]
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5337 entries, 0 to 5616
Data columns (total 13 columns):
CASENO            5337 non-null int64
OFFENSE           5337 non-null object
EVENTDT           5337 non-null object
EVENTTM           5337 non-null object
EVENTDESC         5337 non-null object
D.O.W.            5337 non-null int64
InDbDate          5337 non-null object
Block_Location    5337 non-null object
BLKADDR           5310 non-null object
City              5337 non-null object
State             5337 non-null object
LATITUDE          5337 non-null object
LONGITUDE         5337 non-null object
dtypes: int64(2), object(11)
memory usage: 583.7+ KB


In [60]:
df3 = df[df['LONGITUDE'].str.startswith('-122') & df['LATITUDE'].str.startswith('37')]


In [67]:
df3['LATITUDE'].nunique()

1383

In [77]:
len(df3[['LONGITUDE', 'LATITUDE']].index)

5337

## Drop Columns

We got all the information we needed from "Block_Location" keeping it would be taking up extra room in our dataframe. 

Let's drop the "Block_Location" from the dataframe.

In [ ]:
#df = df.drop("COL_NAME", axis = 1)
df = df.drop("Block_Location", axis = 1)
#Check if it dropped
df.head()

We can drop other columns that we do not think would add useful information to our analysis. 

Although we did get this data from the Berkeley PD, let's make sure all values in "City" are "Berkeley". Also, Let's make sure the "State" is "CA" for all values.

In [ ]:
df.City.unique()

In [ ]:
df['State'].unique()

We checked all the unique values for columns "State" and "City" and they are the results we wanted, therefore, we do not need those columns anymore. 

Drop the columns listed above.

In [ ]:
#drop City and State columns
#df = ...
df = df.drop(["City", "State"], axis = 1)
#Check if they dropped
df.head()

## Dealing With Null Values (NaN)

<img src = "null_def.png">

REFERENCE: https://pandas.pydata.org/pandas-docs/stable/missing_data.html

This is a big data set and we can't look through each value one at a time. How can we make sure that there is a value for each category?

In [78]:
df.isnull().sum()

CASENO             0
OFFENSE            0
EVENTDT            0
EVENTTM            0
EVENTDESC          0
D.O.W.             0
InDbDate           0
Block_Location     0
BLKADDR           27
City               0
State              0
LATITUDE           0
LONGITUDE          0
dtype: int64

Let's look at the rows where "BLKADDR" is a null value. Let's make a temporary sub-dataframe.

In [79]:
null_temp = df[pd.isnull(df['BLKADDR'])]
null_temp

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
53    17036936               DISTURBANCE  06/26/2017 12:00:00 AM   18:24   
104   17090713  THEFT FELONY (OVER $950)  04/09/2017 12:00:00 AM   04:15   
224   17024641             BURGLARY AUTO  05/01/2017 12:00:00 AM   21:00   
235   17046547            VEHICLE STOLEN  08/08/2017 12:00:00 AM   17:00   
291   17053694  THEFT MISD. (UNDER $950)  09/07/2017 12:00:00 AM   17:43   
475   17022572            VEHICLE STOLEN  04/22/2017 12:00:00 AM   21:00   
534   17026854      BURGLARY RESIDENTIAL  05/12/2017 12:00:00 AM   09:00   
1228  17091147             BURGLARY AUTO  06/14/2017 12:00:00 AM   03:00   
1306  17020446            VEHICLE STOLEN  04/12/2017 12:00:00 AM   18:00   
1311  17025351           THEFT FROM AUTO  05/04/2017 12:00:00 AM   22:30   
1550  17029419      BURGLARY RESIDENTIAL  05/23/2017 12:00:00 AM   16:30   
1624  17033574                 NARCOTICS  06/11/2017 12:00:00 AM   16:53   
2093  17037072             BURGLARY AUTO  06/20/2017 12:00:00 AM   08:00   
2340  17028899           ALCOHOL OFFENSE  05/21/2017 12:00:00 AM   22:00   
2595  17030776      BURGLARY RESIDENTIAL  05/30/2017 12:00:00 AM   10:45   
3453  17038382      BURGLARY RESIDENTIAL  06/30/2017 12:00:00 AM   15:00   
3922  17034863               DISTURBANCE  06/01/2017 12:00:00 AM   00:00   
4074  17091632                 VANDALISM  08/15/2017 12:00:00 AM   23:30   
4505  17091900  THEFT MISD. (UNDER $950)  09/20/2017 12:00:00 AM   21:35   
4545  17026051           THEFT FROM AUTO  05/07/2017 12:00:00 AM   16:00   
4759  17091711             BURGLARY AUTO  08/28/2017 12:00:00 AM   08:20   
4852  17042511             BURGLARY AUTO  07/20/2017 12:00:00 AM   16:00   
4959  17090810             BURGLARY AUTO  04/23/2017 12:00:00 AM   00:00   
5032  17030164            VEHICLE STOLEN  05/26/2017 12:00:00 AM   20:00   
5194  17091379             BURGLARY AUTO  07/14/2017 12:00:00 AM   12:00   
5406  17020088     ASSAULT/BATTERY MISD.  04/11/2017 12:00:00 AM   09:25   
5544  17091287                 VANDALISM  07/01/2017 12:00:00 AM   08:00   

                   EVENTDESC  D.O.W.                InDbDate  \
53        DISORDERLY CONDUCT       1  09/25/2017 03:30:16 AM   
104                  LARCENY       0  09/25/2017 03:30:12 AM   
224       BURGLARY - VEHICLE       1  09/25/2017 03:30:12 AM   
235      MOTOR VEHICLE THEFT       2  09/25/2017 03:30:18 AM   
291                  LARCENY       4  09/25/2017 03:30:19 AM   
475      MOTOR VEHICLE THEFT       6  09/25/2017 03:30:12 AM   
534   BURGLARY - RESIDENTIAL       5  09/25/2017 03:30:12 AM   
1228      BURGLARY - VEHICLE       3  09/25/2017 03:30:15 AM   
1306     MOTOR VEHICLE THEFT       3  09/25/2017 03:30:12 AM   
1311  LARCENY - FROM VEHICLE       4  09/25/2017 03:30:12 AM   
1550  BURGLARY - RESIDENTIAL       2  09/25/2017 03:30:13 AM   
1624          DRUG VIOLATION       0  09/25/2017 03:30:14 AM   
2093      BURGLARY - VEHICLE       2  09/25/2017 03:30:16 AM   
2340    LIQUOR LAW VIOLATION       0  09/25/2017 03:30:13 AM   
2595  BURGLARY - RESIDENTIAL       2  09/25/2017 03:30:14 AM   
3453  BURGLARY - RESIDENTIAL       5  09/25/2017 03:30:17 AM   
3922      DISORDERLY CONDUCT       4  09/25/2017 03:30:16 AM   
4074               VANDALISM       2  09/25/2017 03:30:18 AM   
4505                 LARCENY       3  09/25/2017 03:30:19 AM   
4545  LARCENY - FROM VEHICLE       0  09/25/2017 03:30:12 AM   
4759      BURGLARY - VEHICLE       1  09/25/2017 03:30:18 AM   
4852      BURGLARY - VEHICLE       4  09/25/2017 03:30:17 AM   
4959      BURGLARY - VEHICLE       0  09/25/2017 03:30:12 AM   
5032     MOTOR VEHICLE THEFT       5  09/25/2017 03:30:13 AM   
5194      BURGLARY - VEHICLE       5  09/25/2017 03:30:17 AM   
5406                 ASSAULT       2  09/25/2017 03:30:12 AM   
5544               VANDALISM       6  09/25/2017 03:30:17 AM   

                              Block_Location BLKADDR      City State  \
53    Berkeley

Does the number of rows in the dataframe match the values above?

In [80]:
#get number of rows of new df
len(null_temp)


27

Investigate the dataframe, do you see somethng interesting that all these rows share?

Are the Latitude/Longitude values all the same for the "NaN" values?

In [81]:
#get unique values of latitude

null_temp.LATITUDE.unique()

array(['37.869058'], dtype=object)

In [ ]:
#get unique values of longitude

null_temp.LONGITUDE.unique()

## Boolean Slicing

Let's look at the whole dataset to see if there are any rows with that Latitude and Longitude combination that might have a "BLKADDR" associated with it.  

In [82]:
df[(df["LATITUDE"] == '37.869058') & (df["LONGITUDE"] == '-122.270455')]

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
53    17036936               DISTURBANCE  06/26/2017 12:00:00 AM   18:24   
104   17090713  THEFT FELONY (OVER $950)  04/09/2017 12:00:00 AM   04:15   
224   17024641             BURGLARY AUTO  05/01/2017 12:00:00 AM   21:00   
235   17046547            VEHICLE STOLEN  08/08/2017 12:00:00 AM   17:00   
291   17053694  THEFT MISD. (UNDER $950)  09/07/2017 12:00:00 AM   17:43   
475   17022572            VEHICLE STOLEN  04/22/2017 12:00:00 AM   21:00   
534   17026854      BURGLARY RESIDENTIAL  05/12/2017 12:00:00 AM   09:00   
1228  17091147             BURGLARY AUTO  06/14/2017 12:00:00 AM   03:00   
1306  17020446            VEHICLE STOLEN  04/12/2017 12:00:00 AM   18:00   
1311  17025351           THEFT FROM AUTO  05/04/2017 12:00:00 AM   22:30   
1550  17029419      BURGLARY RESIDENTIAL  05/23/2017 12:00:00 AM   16:30   
1624  17033574                 NARCOTICS  06/11/2017 12:00:00 AM   16:53   
2093  17037072             BURGLARY AUTO  06/20/2017 12:00:00 AM   08:00   
2340  17028899           ALCOHOL OFFENSE  05/21/2017 12:00:00 AM   22:00   
2595  17030776      BURGLARY RESIDENTIAL  05/30/2017 12:00:00 AM   10:45   
3453  17038382      BURGLARY RESIDENTIAL  06/30/2017 12:00:00 AM   15:00   
3922  17034863               DISTURBANCE  06/01/2017 12:00:00 AM   00:00   
4074  17091632                 VANDALISM  08/15/2017 12:00:00 AM   23:30   
4505  17091900  THEFT MISD. (UNDER $950)  09/20/2017 12:00:00 AM   21:35   
4545  17026051           THEFT FROM AUTO  05/07/2017 12:00:00 AM   16:00   
4759  17091711             BURGLARY AUTO  08/28/2017 12:00:00 AM   08:20   
4852  17042511             BURGLARY AUTO  07/20/2017 12:00:00 AM   16:00   
4959  17090810             BURGLARY AUTO  04/23/2017 12:00:00 AM   00:00   
5032  17030164            VEHICLE STOLEN  05/26/2017 12:00:00 AM   20:00   
5194  17091379             BURGLARY AUTO  07/14/2017 12:00:00 AM   12:00   
5406  17020088     ASSAULT/BATTERY MISD.  04/11/2017 12:00:00 AM   09:25   
5544  17091287                 VANDALISM  07/01/2017 12:00:00 AM   08:00   

                   EVENTDESC  D.O.W.                InDbDate  \
53        DISORDERLY CONDUCT       1  09/25/2017 03:30:16 AM   
104                  LARCENY       0  09/25/2017 03:30:12 AM   
224       BURGLARY - VEHICLE       1  09/25/2017 03:30:12 AM   
235      MOTOR VEHICLE THEFT       2  09/25/2017 03:30:18 AM   
291                  LARCENY       4  09/25/2017 03:30:19 AM   
475      MOTOR VEHICLE THEFT       6  09/25/2017 03:30:12 AM   
534   BURGLARY - RESIDENTIAL       5  09/25/2017 03:30:12 AM   
1228      BURGLARY - VEHICLE       3  09/25/2017 03:30:15 AM   
1306     MOTOR VEHICLE THEFT       3  09/25/2017 03:30:12 AM   
1311  LARCENY - FROM VEHICLE       4  09/25/2017 03:30:12 AM   
1550  BURGLARY - RESIDENTIAL       2  09/25/2017 03:30:13 AM   
1624          DRUG VIOLATION       0  09/25/2017 03:30:14 AM   
2093      BURGLARY - VEHICLE       2  09/25/2017 03:30:16 AM   
2340    LIQUOR LAW VIOLATION       0  09/25/2017 03:30:13 AM   
2595  BURGLARY - RESIDENTIAL       2  09/25/2017 03:30:14 AM   
3453  BURGLARY - RESIDENTIAL       5  09/25/2017 03:30:17 AM   
3922      DISORDERLY CONDUCT       4  09/25/2017 03:30:16 AM   
4074               VANDALISM       2  09/25/2017 03:30:18 AM   
4505                 LARCENY       3  09/25/2017 03:30:19 AM   
4545  LARCENY - FROM VEHICLE       0  09/25/2017 03:30:12 AM   
4759      BURGLARY - VEHICLE       1  09/25/2017 03:30:18 AM   
4852      BURGLARY - VEHICLE       4  09/25/2017 03:30:17 AM   
4959      BURGLARY - VEHICLE       0  09/25/2017 03:30:12 AM   
5032     MOTOR VEHICLE THEFT       5  09/25/2017 03:30:13 AM   
5194      BURGLARY - VEHICLE       5  09/25/2017 03:30:17 AM   
5406                 ASSAULT       2  09/25/2017 03:30:12 AM   
5544               VANDALISM       6  09/25/2017 03:30:17 AM   

                              Block_Location BLKADDR      City State  \
53    Berkeley

In [ ]:
#get unique values of BLKADDR for the Lat/Long combo

...

## Drop Null Values (NaN)

We could essentially go to Google and try to figure out the BLKADDR ourselves, but to remove any problems that may occur while searching let's just drop all the rows that include null values

In [ ]:
#drop rows that have null values
df = df.dropna(axis = 0, how = "any")

In [ ]:
#now find out how many null values within the df
#What should you see when you run this?
...

# <font color = "red"> YOUR TURN! </font>

### What is the difference between "EVENTDT" and "EVENTTM"? How can we clean our columns to reflect the data that is useful?

HINT: Focus on EVENTDT

In [83]:
#Slice the string to get the information you want and set to the variable
date = [i[:10] for i in df["EVENTDT"]]

#Replace "EVENTDT" with new variable
df["EVENTDT"] = date
df.head()
#check if it worked
df.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

   D.O.W.                InDbDate  \
0       3  09/25/2017 03:30:15 AM   
1       5  09/25/2017 03:30:18 AM   
2       6  09/25/2017 03:30:15 AM   
3       0  09/25/2017 03:30:11 AM   
4       6  09/25/2017 03:30:14 AM   

                                      Block_Location  \
0  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  37.871884  -122.270752  
2               2500 SHATTUCK AVE  Berkeley    CA  37.863811  -122.267412  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  37.867221  -122.263531  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA  37.871884  -122.270752

### Drop the "InDbDate" column

In [84]:
df = df.drop("InDbDate", axis = 1)

#Check that it actually dropped
df.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

   D.O.W.                                     Block_Location  \
0       3  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1       5  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2       6  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3       0  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4       6  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  37.871884  -122.270752  
2               2500 SHATTUCK AVE  Berkeley    CA  37.863811  -122.267412  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  37.867221  -122.263531  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA  37.871884  -122.270752

### TRICKY QUESTION

### Column "D.O.W." can be a bit confusing with the numbers. Replace the numbers with the appropriate day it corresponds to. 

You can find the days it corresponds to in the beginning of the notebook. 

##### HINT: You may need to use a dictionary, the map function, or the zip function

DICT:
https://www.programiz.com/python-programming/methods/built-in/dict

MAP:
https://www.programiz.com/python-programming/methods/built-in/map

ZIP:
https://www.programiz.com/python-programming/methods/built-in/zip

In [85]:
dow = {0:"Sunday",1:"Monday",2:"Tuesday",3:"Wednesday", 4:"Thursday", 5:"Friday", 6:"Saturday"}
df["D.O.W."] = df["D.O.W."].map(dow)

#Check if it worked
df.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

      D.O.W.                                     Block_Location  \
0  Wednesday  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1     Friday  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2   Saturday  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3     Sunday  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4   Saturday  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  37.871884  -122.270752  
2               2500 SHATTUCK AVE  Berkeley    CA  37.863811  -122.267412  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  37.867221  -122.263531  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA  37.871884  -122.270752

# Part 2: EXPLORATORY DATA ANALYSIS

<h3>"Exploratory data analysis or 'EDA' is a <b>critical</b> beginning step in analyzing the data from an experiment.</h3>

<b>Here are the main reasons we use EDA:</b>
<ul>
• detection of mistakes<br><br>
• checking of assumptions<br><br>
• preliminary selection of appropriate models<br><br>
• determining relationships among the explanatory variables, and<br><br>
• assessing the direction and rough size of relationships between explanatory and outcome variables."</ul>
REFERENCE: http://www.stat.cmu.edu/~hseltman/309/Book/chapter4.pdf


## Now what?

We have cleaned our data to the best of our ability based on the initial look. Now let's try to look at the <b>relationships</b> between different values. 

In [86]:
df.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

      D.O.W.                                     Block_Location  \
0  Wednesday  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1     Friday  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2   Saturday  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3     Sunday  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4   Saturday  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  37.871884  -122.270752  
2               2500 SHATTUCK AVE  Berkeley    CA  37.863811  -122.267412  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  37.867221  -122.263531  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA  37.871884  -122.270752

Let's look at the different types of offenses that were called in. We know that using the .unique() function will return all the unique values in the column, but what if we wanted to also <b>count</b> the different times each unique value appeared?

In [ ]:
df.OFFENSE.value_counts()

In [ ]:
df.EVENTDESC.value_counts()

Why is "LARCENY" a higher occurence in the "EVENTDESC" column, if when we looked into the "OFFENSE" column, "BURGLARY - VEHICLE" is first? Let's look into this a little more.


## GroupBy 

In [ ]:
df.groupby("EVENTDESC").OFFENSE.value_counts()

#turn the series into a DF 
df.groupby("EVENTDESC").OFFENSE.value_counts().to_frame()

From this DF we can see that "LARCENY" has the most OFFENSES within that category. When you add the totals from "THEFT MISD ( UNDER $ 950)" ," THEFT  FELONY (OVER $ 950)", and
"THEFT FROM PERSON" they equal more than "BURGLARY - VEHICLE", but "BURGLARY AUTO" as an offense alone is the highest in number.

# <font color = "red"> YOUR TURN! </font>

Could there be any relationship with the Day of the Week and the calls? Try out different functions to see if there is any significance?

In [87]:
#count the amount of calls per day
df["D.O.W."].value_counts()

Tuesday      859
Friday       822
Wednesday    821
Saturday     811
Thursday     801
Monday       796
Sunday       707
Name: D.O.W., dtype: int64

With the day that has the most calls, check the type of offense that appears the most.

In [ ]:
#only display rows with the D.O.W that appears the most
#create a temp df
...

In [ ]:
#count the number of offenses by type
...

In [ ]:
#count the number of eventdesc by type
...

Do these numbers match the results of the overall DF?


Let's try something else, Friday and Saturday nights are typically associated with being the "party" time. If this is true should there be more Liquor/Drug/Disordely Conduct/etc. occurrences those nights?

Let's try it with Fridays!

In [116]:
#only display rows with the D.O.W of Friday
#create a temp df

fri =df[df["D.O.W."] == "Friday"]
fri.head()

CASENO                   OFFENSE     EVENTDT EVENTTM  \
1   17052235                GUN/WEAPON  09/01/2017   22:56   
5   17043981           ALCOHOL OFFENSE  07/28/2017   21:31   
6   17091423  THEFT FELONY (OVER $950)  07/21/2017   20:35   
14  17090740            IDENTITY THEFT  04/14/2017   12:55   
16  17048522     ASSAULT/BATTERY MISD.  08/18/2017   07:45   

               EVENTDESC  D.O.W.  \
1        WEAPONS OFFENSE  Friday   
5   LIQUOR LAW VIOLATION  Friday   
6                LARCENY  Friday   
14                 FRAUD  Friday   
16               ASSAULT  Friday   

                                       Block_Location  \
1   UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
5   KITTREDGE STREET &amp; FULTON ST\nBerkeley, CA...   
6   2400 HASTE ST\nBerkeley, CA\n(37.865849, -122....   
14  900 SAN BENITO RD\nBerkeley, CA\n(37.892152, -...   
16  2100 OXFORD ST\nBerkeley, CA\n(37.87091, -122....   

                          BLKADDR      City State   LATITUDE    LONGITUDE  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  37.871884  -122.270752  
5    KITTREDGE STREET & FULTON ST  Berkeley    CA  37.868706  -122.266279  
6                   2400 HASTE ST  Berkeley    CA  37.865849  -122.259977  
14              900 SAN BENITO RD  Berkeley    CA  37.892152  -122.269211  
16                 2100 OXFORD ST  Berkeley    CA   37.87091  -122.265993

In [ ]:
#count the number of offenses by type
...

In [ ]:
#count the number of events by type
...

This will not give us the information we want. Instead let us look at each EVENTDESC and group by the D.O.W. that appears the most per EVENTDESC.

Also, I realized that "D.O.W." is becoming a problem with the periods when I am trying to call my series. I want to change the name of the column again, how can I do that?

In [ ]:
#change column name
df = ...
df.head()

# <font color = "red"> GROUP WORK</font> 
## What do YOU want to find out? YOUR DATA INVESTIGATION

In this notebook you have been learning all these techniques to be able to manipulate your dataframe to your preference. We know how to clean and explore our data, but what questions or topics did you actually want to learn from the data? 

<b> * In groups of 2-4 people, investigate the dataframe in this notebook and pick a question/topic to answer. Using the techniques you learned today, show relationships and results that would support that question/topic. 
</b><br><br>
<i>If we have time</i> <b>each</b> group will present their investigations and why they are significant to the class.


In [146]:
df3 = df[df['LONGITUDE'].str.startswith('-122') & df['LATITUDE'].str.startswith('37')]
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5337 entries, 0 to 5616
Data columns (total 12 columns):
CASENO            5337 non-null int64
OFFENSE           5337 non-null object
EVENTDT           5337 non-null object
EVENTTM           5337 non-null object
EVENTDESC         5337 non-null object
D.O.W.            5337 non-null object
Block_Location    5337 non-null object
BLKADDR           5310 non-null object
City              5337 non-null object
State             5337 non-null object
LATITUDE          5337 non-null object
LONGITUDE         5337 non-null object
dtypes: int64(1), object(11)
memory usage: 542.0+ KB


In [147]:
df3['LONGITUDE'] = pd.to_numeric(df3['LONGITUDE'])
df3['LATITUDE'] = pd.to_numeric(df3['LATITUDE'])


/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [175]:
import numpy
df3['LATITUDE'] = ((df3['LATITUDE'] + 0.005) * 100).astype(numpy.int64)/100

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [176]:
import numpy
df3['LONGITUDE'] = ((df3['LONGITUDE'] + 0.005) * 100).astype(numpy.int64)/100


/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [179]:
df3.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

      D.O.W.                                     Block_Location  \
0  Wednesday  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1     Friday  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2   Saturday  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3     Sunday  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4   Saturday  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State  LATITUDE  LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA     37.87    -122.26  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA     37.87    -122.26  
2               2500 SHATTUCK AVE  Berkeley    CA     37.86    -122.26  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA     37.87    -122.25  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA     37.87    -122.26

In [218]:
df4 = df3.groupby(['LATITUDE','LONGITUDE']).size().reset_index()
df5 = df4.rename(columns = {0:'OCCURRENCE'})
df5

LATITUDE  LONGITUDE  OCCURRENCE
0      37.82    -122.27           6
1      37.85    -122.29           9
2      37.85    -122.28         139
3      37.85    -122.27         152
4      37.85    -122.26         138
5      37.85    -122.25          26
6      37.85    -122.24          34
7      37.85    -122.23           7
8      37.86    -122.31          55
9      37.86    -122.29          84
10     37.86    -122.28         196
11     37.86    -122.27         157
12     37.86    -122.26         339
13     37.86    -122.25         291
14     37.86    -122.24         143
15     37.86    -122.23          23
16     37.86    -122.22           1
17     37.87    -122.31           6
18     37.87    -122.30          87
19     37.87    -122.29         307
20     37.87    -122.28         283
21     37.87    -122.27         208
22     37.87    -122.26         918
23     37.87    -122.25         462
24     37.87    -122.24         179
25     37.87    -122.23           4
26     37.88    -122.30          30
27     37.88    -122.29         263
28     37.88    -122.28          91
29     37.88    -122.27          88
30     37.88    -122.26         216
31     37.88    -122.25         100
32     37.88    -122.24          16
33     37.89    -122.30           2
34     37.89    -122.28          10
35     37.89    -122.27          49
36     37.89    -122.26          79
37     37.89    -122.25          37
38     37.89    -122.24          20
39     37.90    -122.28           4
40     37.90    -122.27          20
41     37.90    -122.26          35
42     37.90    -122.25          22
43     37.91    -122.27           1

In [221]:
df5[df5.OCCURRENCE == df5.OCCURRENCE.max()]

LATITUDE  LONGITUDE  OCCURRENCE
22     37.87    -122.26         918

In [283]:
(df5[df5.OCCURRENCE == df5.OCCURRENCE.max()])[['LONGITUDE', 'LATITUDE','OCCURRENCE']]

LONGITUDE  LATITUDE  OCCURRENCE
22    -122.26     37.87         918

In [284]:
df6 = df3[ df3['LONGITUDE'].isin((df5[df5.OCCURRENCE == df5.OCCURRENCE.max()]['LONGITUDE'])) & df3['LATITUDE'].isin((df5[df5.OCCURRENCE == df5.OCCURRENCE.max()]['LATITUDE']))   ]

In [295]:
df6['D.O.W.'].value_counts()

Tuesday      149
Saturday     132
Friday       132
Monday       131
Thursday     129
Wednesday    127
Sunday       118
Name: D.O.W., dtype: int64

In [296]:
df6[df6['D.O.W.'] == df6['D.O.W.'].max()]['EVENTTM'].value_counts()

08:00    6
15:00    4
23:00    4
12:00    3
13:00    3
16:00    3
11:00    2
18:00    2
20:55    2
11:30    2
17:55    2
15:46    2
20:56    2
21:00    2
02:30    2
22:00    2
00:00    2
15:10    1
17:00    1
14:00    1
11:45    1
09:56    1
04:04    1
21:37    1
08:55    1
09:06    1
07:47    1
19:25    1
08:50    1
03:00    1
        ..
20:40    1
15:50    1
16:16    1
18:10    1
08:40    1
21:47    1
18:19    1
21:35    1
01:22    1
07:55    1
10:00    1
15:25    1
14:53    1
15:45    1
00:30    1
08:20    1
06:00    1
13:23    1
01:00    1
18:30    1
01:50    1
00:35    1
19:30    1
22:10    1
06:15    1
13:01    1
16:45    1
22:52    1
13:50    1
23:18    1
Name: EVENTTM, Length: 99, dtype: int64

In [297]:
df6[df6['D.O.W.'] == df6['D.O.W.'].max()]['EVENTDESC'].value_counts()

LARCENY                   32
BURGLARY - VEHICLE        20
DISORDERLY CONDUCT        18
ASSAULT                   11
DRUG VIOLATION             9
VANDALISM                  8
ROBBERY                    6
MOTOR VEHICLE THEFT        5
BURGLARY - COMMERCIAL      3
FRAUD                      3
BURGLARY - RESIDENTIAL     3
WEAPONS OFFENSE            2
LIQUOR LAW VIOLATION       2
MISSING PERSON             2
FAMILY OFFENSE             2
LARCENY - FROM VEHICLE     1
Name: EVENTDESC, dtype: int64

In [309]:
df6.head()

CASENO                OFFENSE     EVENTDT EVENTTM             EVENTDESC  \
0   17034187          BURGLARY AUTO  06/14/2017   15:15    BURGLARY - VEHICLE   
1   17052235             GUN/WEAPON  09/01/2017   22:56       WEAPONS OFFENSE   
4   17033328              NARCOTICS  06/10/2017   14:30        DRUG VIOLATION   
5   17043981        ALCOHOL OFFENSE  07/28/2017   21:31  LIQUOR LAW VIOLATION   
16  17048522  ASSAULT/BATTERY MISD.  08/18/2017   07:45               ASSAULT   

       D.O.W.                                     Block_Location  \
0   Wednesday  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1      Friday  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
4    Saturday  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   
5      Friday  KITTREDGE STREET &amp; FULTON ST\nBerkeley, CA...   
16     Friday  2100 OXFORD ST\nBerkeley, CA\n(37.87091, -122....   

                           BLKADDR      City State  LATITUDE  LONGITUDE  
0       ALLSTON WAY & SHATTUCK AVE  Berkeley    CA     37.87    -122.26  
1    UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA     37.87    -122.26  
4   MILVIA STREET & UNIVERSITY AVE  Berkeley    CA     37.87    -122.26  
5     KITTREDGE STREET & FULTON ST  Berkeley    CA     37.87    -122.26  
16                  2100 OXFORD ST  Berkeley    CA     37.87    -122.26

In [325]:
def group_by_4_hour(x):
    hour, _ = x['EVENTTM'].split(':')
    if int(hour) < 4 : return ('0:00 - 4:00')
    if int(hour) < 8 : return ('4:00 - 8:00')
    if int(hour) < 12 : return ('8:00 - 12:00')
    if int(hour) < 16 : return ('12:00 - 16:00')
    if int(hour) < 20 : return ('16:00 - 20:00')
    return ('20:00 - 24:00')
#period = [[i for i in ] for j in x['EVENTDT']]
df6['PERIOD'] = df6.apply(group_by_4_hour, axis=1)

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # Remove the CWD from sys.path while we load stuff.


In [331]:
df6.head()

CASENO                OFFENSE     EVENTDT EVENTTM             EVENTDESC  \
0   17034187          BURGLARY AUTO  06/14/2017   15:15    BURGLARY - VEHICLE   
1   17052235             GUN/WEAPON  09/01/2017   22:56       WEAPONS OFFENSE   
4   17033328              NARCOTICS  06/10/2017   14:30        DRUG VIOLATION   
5   17043981        ALCOHOL OFFENSE  07/28/2017   21:31  LIQUOR LAW VIOLATION   
16  17048522  ASSAULT/BATTERY MISD.  08/18/2017   07:45               ASSAULT   

       D.O.W.                                     Block_Location  \
0   Wednesday  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1      Friday  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
4    Saturday  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   
5      Friday  KITTREDGE STREET &amp; FULTON ST\nBerkeley, CA...   
16     Friday  2100 OXFORD ST\nBerkeley, CA\n(37.87091, -122....   

                           BLKADDR      City State  LATITUDE  LONGITUDE  \
0       ALLSTON WAY & SHATTUCK AVE  Berkeley    CA     37.87    -122.26   
1    UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA     37.87    -122.26   
4   MILVIA STREET & UNIVERSITY AVE  Berkeley    CA     37.87    -122.26   
5     KITTREDGE STREET & FULTON ST  Berkeley    CA     37.87    -122.26   
16                  2100 OXFORD ST  Berkeley    CA     37.87    -122.26   

           PERIOD  
0   12:00 - 16:00  
1   20:00 - 24:00  
4   12:00 - 16:00  
5   20:00 - 24:00  
16    4:00 - 8:00

In [333]:
df6.PERIOD.max()

'8:00 - 12:00'

In [337]:
df6[df6.PERIOD == df6.PERIOD.max()].groupby(['EVENTDESC']).count()

CASENO  OFFENSE  EVENTDT  EVENTTM  D.O.W.  \
EVENTDESC                                                           
ALL OTHER OFFENSES           1        1        1        1       1   
ARSON                        1        1        1        1       1   
ASSAULT                     21       21       21       21      21   
BURGLARY - COMMERCIAL        1        1        1        1       1   
BURGLARY - RESIDENTIAL       8        8        8        8       8   
BURGLARY - VEHICLE          26       26       26       26      26   
DISORDERLY CONDUCT          16       16       16       16      16   
DRUG VIOLATION               3        3        3        3       3   
FAMILY OFFENSE               2        2        2        2       2   
FRAUD                        2        2        2        2       2   
LARCENY                     55       55       55       55      55   
LARCENY - FROM VEHICLE       2        2        2        2       2   
LIQUOR LAW VIOLATION         1        1        1        1       1   
MISSING PERSON               4        4        4        4       4   
MOTOR VEHICLE THEFT          3        3        3        3       3   
ROBBERY                      6        6        6        6       6   
VANDALISM                   18       18       18       18      18   
WEAPONS OFFENSE              1        1        1        1       1   

                        Block_Location  BLKADDR  City  State  LATITUDE  \
EVENTDESC                                                                
ALL OTHER OFFENSES                   1        1     1      1         1   
ARSON                                1        1     1      1         1   
ASSAULT                             21       20    21     21        21   
BURGLARY - COMMERCIAL                1        1     1      1         1   
BURGLARY - RESIDENTIAL               8        6     8      8         8   
BURGLARY - VEHICLE                  26       24    26     26        26   
DISORDERLY CONDUCT                  16       16    16     16        16   
DRUG VIOLATION                       3        3     3      3         3   
FAMILY OFFENSE                       2        2     2      2         2   
FRAUD                                2        2     2      2         2   
LARCENY                             55       55    55     55        55   
LARCENY - FROM VEHICLE               2        2     2      2         2   
LIQUOR LAW VIOLATION                 1        1     1      1         1   
MISSING PERSON                       4        4     4      4         4   
MOTOR VEHICLE THEFT                  3        3     3      3         3   
ROBBERY                              6        6     6      6         6   
VANDALISM                           18       17    18     18        18   
WEAPONS OFFENSE                      1        1     1      1         1   

                        LONGITUDE  PERIOD  
EVENTDESC                                  
ALL OTHER OFFENSES              1       1  
ARSON                           1       1  
ASSAULT                        21      21  
BURGLARY - COMMERCIAL           1       1  
BURGLARY - RESIDENTIAL          8       8  
BURGLARY - VEHICLE             26      26  
DISORDERLY CONDUCT             16      16  
DRUG VIOLATION                  3       3  
FAMILY OFFENSE                  2       2  
FRAUD                           2       2  
LARCENY                        55      55  
LARCENY - FROM VEHICLE          2       2  
LIQUOR LAW VIOLATION            1       1  
MISSING PERSON                  4       4  
MOTOR VEHICLE THEFT             3       3  
ROBBERY                         6       6  
VANDALISM                      18      18  
WEAPONS OFFENSE                 1       1